### ISSUE TITLE
Segfault using `torch.unique` on tensor with NaNs with `dim=0` on PyTorch 1.10

### ISSUE THREAD

**Issue by jan-matthis** [pytorch/pytorch#71089]
### 🐛 Describe the bug

`torch.unique(torch.tensor([float("nan")]), dim=0)` yields a segfault on PyTorch 1.10 while working fine on 1.9.

### To reproduce the issue

```python
# segfaults in torch 1.10, works as expected in torch 1.9:
tensor_nan = torch.tensor([float("nan")])
assert torch.unique(tensor_nan, dim=0).numel() == 1
assert torch.isnan(torch.unique(tensor_nan, dim=0)[0])
```

### Additional comments

Note that when leaving `dim` to its default (`None`), the segfault does not occur, i.e.:
```python
# works as expected in both torch 1.10 and 1.9:
tensor_nan = torch.tensor([float("nan")])
assert torch.unique(tensor_nan, dim=None).numel() == 1
assert torch.isnan(torch.unique(tensor_nan, dim=None)[0])
```

The problem does not occur for `float("inf")`, i.e.:
```python
# works as expected in both torch 1.10 and 1.9:
tensor_inf = torch.tensor([float("inf")])
assert torch.unique(tensor_inf, dim=0).numel() == 1
assert torch.isinf(torch.unique(tensor_inf, dim=0)[0])
```

### Versions

```
PyTorch version: 1.10.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.0.1 (x86_64)
GCC version: Could not collect
Clang version: 13.0.0 (clang-1300.0.29.30)
CMake version: Could not collect
Libc version: N/A

Python version: 3.8.12 | packaged by conda-forge | (default, Oct 12 2021, 21:50:56)  [Clang 11.1.0 ] (64-bit runtime)
Python platform: macOS-10.16-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] torch==1.10.1
[conda] torch                     1.10.1                   pypi_0    pypi
```

cc @ezyang @gchanan @zou3519 @bdhirsh

**Comment by jbschlosser** [comment]
Reproduced the segfault on master

**Comment by dagitses** [comment]
High priority due to crash and regression.

**Comment by ngimel** [comment]
Likely caused by #64835, `at::equal` returns `false` for comparing 2 nans (but `inf` should be fine)

**Comment by oToToT** [comment]
Yes, probably there's problem there, and we might also have a similar problem in `torch.unique_consecutive`.

I test the function `torch.unique_consecutive` on [colab](https://colab.research.google.com/drive/1CyJQcWu44gL2HtEH1_6XF_HzRp7HxdZr?usp=sharing), and it also seems not working properly.
